config/config.go

```diff
@@ -186,19 +196,50 @@ func FromViper(v *viper.Viper) (*Config, error) {
 		cfg.Walk = walk.Stdin.String()
 	}
 
+	// set git-based tree root command if the walker is Git
+	if cfg.Walk == walk.Git.String() {
+		cfg.TreeRootCmd = "git rev-parse --show-toplevel"
+	}
+
 	// determine the tree root
-	if cfg.TreeRoot == "" {
-		// if none was specified, we first try with tree-root-file
-		if cfg.TreeRootFile != "" {
-			// search the tree root using the --tree-root-file if specified
-			_, cfg.TreeRoot, err = FindUp(cfg.WorkingDirectory, cfg.TreeRootFile)
-			if err != nil {
-				return nil, fmt.Errorf("failed to find tree-root based on tree-root-file: %w", err)
-			}
-		} else {
-			// otherwise fallback to the directory containing the config file
-			cfg.TreeRoot = filepath.Dir(v.ConfigFileUsed())
+	// NOTE: at the treefmt command level we ensure that `--tree-root`, `--tree-root-cmd` and `--tree-root-file` are
+	// mutually exclusive
+	switch {
+	case cfg.TreeRoot == "" && cfg.TreeRootFile != "":
+		// search the tree root using the --tree-root-file if specified
+		_, cfg.TreeRoot, err = FindUp(cfg.WorkingDirectory, cfg.TreeRootFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find tree-root based on tree-root-file: %w", err)
+		}
+
+	case cfg.TreeRoot == "" && cfg.TreeRootCmd != "":
+		// use the output of the tree root command as the tree root
+		parts, splitErr := shlex.Split(cfg.TreeRootCmd)
```

Comment:
Should the use of shlex to parse the cmd be documented somewhere?

Alternatively, would it make sense to use `bash` to evaluate the command (`bash -c "$cmd"`)?

Using a "real" shell to eval the command may be more intuitive for users; e.g. they can use things like `$PWD` or `$(pwd)` within their command string. In the future, treefmt could also set its own env vars that the command would have access to.

Another alternative would be to allow `tree-root-cmd` to be either a `str` or `str[]`. Rather than parsing `'` syntax, multiple args could be separated as list elements. I guess this is only useful in toml though, not env-vars or cli-flags.

Reply:
I agree with @Mic92 on this one.